initrd/bin/usb-parse-boot

```diff
@@ -0,0 +1,171 @@
+#!/bin/sh
+file=$1
+basefile=$(basename $file)
+
+reset_entry() {
+	name=""
+	kexectype="elf"
+	kernel=""
+	initrd=""
+	modules=""
+	append=""
+}
+
+echo_entry() {
+	if [ "$kexectype" = "elf" ]; then
+		if [ -z "$kernel" -o -z "$initrd" ]; then return; fi
+
+		entry="$name|$kexectype|kernel $kernel|initrd $initrd"
+		if [ -n "$append" ]; then entry="$entry|append $append"; fi
+		echo "$entry"
+	fi
+	if [ "$kexectype" = "multiboot" ]; then
+		if [ -z "$kernel" ]; then return; fi
+
+		echo "$name|$kexectype|kernel $kernel$modules"
+	fi
+}
+
+search_entry() {
+	case $line in
+		menuentry* | MENUENTRY* )
+			state="grub"
+			reset_entry
+			name=`echo $line | tr "'" "\"" | cut -d\" -f 2`
+			;;
+
+		label* | LABEL* )
+			state="syslinux"
+			reset_entry
+	esac
+}
+
+grub_entry() {
+	if [ "$line" = "}" ]; then
+		echo_entry
+		state="search"
+		return
+	fi
+
+	# add info to menuentry
+	trimcmd=`echo $line | tr '\t ' ' ' | tr -s ' '`
+	cmd=`echo $trimcmd | cut -d\  -f1`
+	val=`echo $trimcmd | cut -d\  -f2-`
+	case $cmd in
+		multiboot*)
+			kexectype="multiboot"
+			kernel="$val"
+			;;
+		module*)
+			modules="$modules|module $val"
+			;;
+		linux*)
+			kernel=`echo $trimcmd | cut -d\  -f2`
+			append=`echo $trimcmd | cut -d\  -f3-`
+			;;
+		initrd*)
+			initrd="$val"
+			;;
+	esac
+}
+
+syslinux_end() {
+	# finish menuentry
+
+	# attempt to parse out of append if missing initrd
+	if [ -z "$initrd" ]; then
+		newappend=""
+		for param in $append; do
+			case $param in
+				initrd=*)
+					initrd=`echo $param | cut -d\= -f2`
+					;;
+				*) newappend="$newappend $param" ;;
+			esac
+		done
+		append="$newappend"
+	fi
+
+	echo_entry
+	state="search"
+}
+
+syslinux_multiboot_append() {
+	splitval=`echo "${val// --- /|}" | tr '|' '\n'`
+	while read line
+	do
+		if [ -z "$kernel" ]; then
+			kernel="$line"
+		else
+			modules="$modules|module $line"
+		fi
+	done << EOF
+$splitval
+EOF
+}
+
+syslinux_entry() {
+	label_line=${line:0:5}
+	if [ -z "$line" ]; then
+		syslinux_end
+	fi
+	if [ "$label_line" = "label" -o "$label_line" = "LABEL" ]; then
+		search_entry
+		return
+	fi
+
+	# add info to menuentry
+	trimcmd=`echo $line | tr '\t ' ' ' | tr -s ' '`
+	cmd=`echo $trimcmd | cut -d\  -f1`
+	val=`echo $trimcmd | cut -d\  -f2-`
+	case $trimcmd in
+		menu* | MENU* )
+			cmd2=`echo $trimcmd | cut -d \  -f2`
+			if [ "$cmd2" = "label" -o "$cmd2" = "LABEL" ]; then
+				name=`echo ${trimcmd:11} | tr -d '^'`
+			fi
+			;;
+		linux* | LINUX* | kernel* | KERNEL* )
+			case $val in
+				*mboot.c32) kexectype="multiboot" ;;
+				*.c32)
+					# skip this entry
+					state="search"
+					;;
+				*)
+					kernel="$val"
+			esac
+			;;
+		initrd* | INITRD* )
+			initrd="$val"
+			;;
+		append* | APPEND* )
+			if [ "$kexectype" = "multiboot" ]; then
+				syslinux_multiboot_append
+			else
+				append="$val"
+			fi
+			;;
+	esac
+}
+
+state="search"
+while read line
```

Comment:
Definitely not the best we can do, especially for fixed installations like the normal Qubes installation.  My goal was to get some external options working first (including cases where grub might not exist), so this simple parsing was just quick and dirty.

I've looked into replicating the grub environment/commands necessary to make at least the qubes install/update/boot process a little more generic and more extensible later (support loopback command for iso etc.).  It should be straight forward to pull out the boot devices and kexec params (if we're focusing Qubes), but I think there will still be special handling for the secret management and the xen binary replacement.

Did you want me to update this PR for this or start another one to try to rework the fixed media / normal Qubes procedure to require less user input?  If there are other fixed OS setups that you're targeting to handle, let me know and I can take a look at the boot routines too.